pkg/canary/writer/push.go

```diff
@@ -0,0 +1,208 @@
+package writer
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"crypto/tls"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/grafana/loki/pkg/util/build"
+)
+
+const (
+	defaultContentType         = "application/x-protobuf"
+	defaultMaxReponseBufferLen = 1024
+
+	pushEndpoint = "/loki/api/v1/push"
+)
+
+var (
+	defaultUserAgent = fmt.Sprintf("canary-push/%s", build.GetVersion().Version)
+)
+
+// Push is a io.Writer, that writes given long entries by pushing
+// directly to the given loki server URL. Each `Push` instance handles for a single tenant.
+// No batching of log lines happens when sending to Loki.
```

Comment:
Why no batching?

Reply:
The main rationale is two folded

1. Handling batch means, we need to take care of the lifecycle of the pusher during shutdown, say flush the batch before stopping. Kinda complicates the existing canary model a bit.
2.Given we push 1 or 2 logs (with just tiny payload) per sec. Didn't wanted to add these complexity now.